### PR TITLE
Add Carthage support

### PR DESCRIPTION
**Pull request by jakecraige**
Added a `--carthage` option that's disabled by default to provide backwards
compatibility.

What this option does:

- Adds an empty `Cartfile`
- Runs `carthage update`
- Adds the `copy-frameworks.sh` build phase to the project
- Adds `carthage update` to `bin/setup` script

### CODE REVIEW

**Comment by jakecraige** [comment]
Right now the `Cartfile` is empty, I'm curious what y'alls thoughts are on providing defaults? If we want to, it's probably best in another PR since it increases the complexity a bit to do all the setup necessary.

**Comment by jakecraige** [comment]
@thoughtbot/ios Any thoughts on this?

**Comment by gfontenot** [comment]
Oh, yeah, sorry. I do have thoughts.

I _really_ want Carthage support. It bothers me that we don't have it already. But I'm also thinking that the better long term solution would be to remove explicit CocoaPods support, and instead make dependency management completely generic. So for example, we could provide a `dependency_installation_command` or something to the `.liftoffrc` and then have it default to `pod install`. That could be overridden to be `carthage bootstrap` or `bin/setup` or `git submodule update --init --recursive` or whatever you want. Maybe we also include a default `dependency_template` or something that defaults to `Podfile`, then we can override that with `Cartfile` or `false` if there isn't one.

This might (will) cause issues with some current implementations that explicitly look for cocoapods support. For example, the Settings bundle stuff is fairly tightly coupled to the current implementation. But I do think this is a better long-term solution.

**Comment by jakecraige** [comment]
@gfontenot I was thinking it would be nice to not duplicate all this logic as I was writing it but didn't want to change everything.

As a middle ground between the current state of this and what you're thinking, what do you think about _for now_ extracting the dependency stuff into a single class so we remove this duplication, and come back later and couple it less?

**Comment by gfontenot** [comment]
Sounds like a reasonable plan. I realize that what I'm asking for is a much larger undertaking.

File: lib/liftoff/cli.rb [hunk 1]
```diff
@@ -35,8 +35,18 @@ def global_options
           @options[:strict_prompts] = strict_prompts
         end
 
+        opts.on('--dependency-managers [NAME(s)]', 'Comma separated list of dependency managers to enable. Available options: cocoapods') do |list|
+          @options[:dependency_managers] = (list || "").split(",")
+        end
+
         opts.on('--[no-]cocoapods', 'Enable/Disable Cocoapods') do |use_cocoapods|
```
Comment (jakecraige) [inline lib/liftoff/cli.rb]: I wasn't sure if I should remove this option or not. It wouldn't break people's `liftoffrc` but if there were a script that passes in options this needs to be here. Thoughts?

File: lib/liftoff/cli.rb [hunk 2]
```diff
@@ -35,8 +35,18 @@ def global_options
           @options[:strict_prompts] = strict_prompts
         end
 
+        opts.on('--dependency-managers [NAME(s)]', 'Comma separated list of dependency managers to enable. Available options: cocoapods,carthage') do |list|
+          @options[:dependency_managers] = (list || "").split(",")
+        end
+
         opts.on('--[no-]cocoapods', 'Enable/Disable Cocoapods') do |use_cocoapods|
```
Comment (jakecraige) [inline lib/liftoff/cli.rb]: I wasn't sure if I should remove this option or not. It wouldn't break people's liftoffrc but if there were a script that passes in options this needs to be here. Thoughts?

**Comment by jakecraige** [comment]
@gfontenot Did the extraction 8625e6e703362d8c270b97bd4555188f8c9135bc, let me know what you think.

File: lib/liftoff/launchpad.rb [hunk 3]
```diff
@@ -13,10 +13,10 @@ def liftoff(options)
 
         file_manager.create_project_dir(@config.path) do
           generate_project
-          setup_cocoapods
+          setup_dependency_managers
           generate_templates
           generate_settings
-          install_cocoapods
+          install_dependency_managers
```
Comment (gfontenot) [inline lib/liftoff/launchpad.rb]: maybe `install_dependencies`?

**Comment by gfontenot** [comment]
This is really really great. I think for the Cartfile, we can at least add `Quick` and `Nimble` as default dependencies. That's in line with what we're doing with CocoaPods, and we'll want them in all of our projects.

**Comment by jakecraige** [comment]
@gfontenot Unfortunately that adds a lot of extra complexity to this PR. With that in mind I'd prefer to keep away from that for now.

**Comment by gfontenot** [comment]
:dancers:

**Pull request merged**